File: Assets/Lua/Rifle.lua
local Rifle = {
    maxMagazineSize = 30, --最大弹容量
    duration = 0.4, --枪口复位时间
    fireInterval = 0.07, --开火间隔时间
    maxAngleDeg = 2, --最大散射角度
    
    startAngle = 0, --当前后坐力角度
    targetOffset = 2 --后坐力威力
}

function Rifle.GetRecoil(elapsed)

    local t = elapsed / Rifle.duration

    if t < 0.3 then --上升
        local recoilT = t / 0.3
        local smoothT = recoilT * recoilT * (3 - 2 * recoilT)
        return CS.UnityEngine.Mathf.Lerp(Rifle.startAngle, Rifle.startAngle - Rifle.targetOffset, smoothT)
    else --下降
        local recoilT = (t - 0.3) / 0.7
        local smoothT = recoilT * recoilT * (3 - 2 * recoilT)
        return CS.UnityEngine.Mathf.Lerp(Rifle.startAngle - Rifle.targetOffset, Rifle.startAngle, smoothT)
    end

end

return Rifle

File: Assets/Scripts/Weapon/Rifle.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

public class Rifle : Gun

{
    protected override void Start()
    {
        base.Start();

        gunType = GunType.Long;

        DataInit();
    }

    protected override void Update()
    {
        base.Update();
    }

    //从Lua脚本获取数据
    protected override void DataInit()
    {
        maxMagazineSize = LuaManager.instance.rifleTab.Get<int>("maxMagazineSize");
        duration = LuaManager.instance.rifleTab.Get<float>("duration");
        maxAngleDeg = LuaManager.instance.rifleTab.Get<float>("maxAngleDeg");
    }

    public override bool Fire(Player player, float moveSpeed, int fireCount)
    {
        if (currentMagazineSize > 0 && fireTime > LuaManager.instance.rifleTab.Get<float>("fireInterval"))
        {
            shootController.Fire(player, moveSpeed, fireCount, muzzleTransform, cam, maxAngleDeg);

            currentMagazineSize--;
            fireTime = 0;

            return true;
        }
        else
        {
            Debug.Log("弹夹为空||开火间隔");

            return false;
        }
    }

    public override void Reload(bool isReloading)
    {
        base.Reload(isReloading);
    }

    //获取枪支后坐力
    public override float GetGunRecoil(float elapsed)
    {
        return LuaManager.instance.getHandGunRecoil(elapsed);
    }
}
